src/google/protobuf/descriptor.cc

```diff
@@ -5480,6 +5490,88 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  bool is_proto2 = result->file()->syntax() == FileDescriptor::SYNTAX_PROTO2;
+  std::string message_name = result->full_name();
+  // two passes: one looking only at default JSON names, and one that considers
+  // custom JSON names
+  CheckFieldJsonNameUniqueness(message_name, proto, is_proto2, false);
+  CheckFieldJsonNameUniqueness(message_name, proto, is_proto2, true);
+}
+
+namespace {
+// Helpers for function below
+
+std::string GetJsonName(const FieldDescriptorProto& field, bool use_custom,
+                        bool* was_custom) {
+  if (use_custom && field.has_json_name()) {
+    *was_custom = true;
+    return field.json_name();
+  }
+  *was_custom = false;
+  return ToJsonName(field.name());
+}
+
+struct JsonNameDetails {
+  const FieldDescriptorProto* field;
+  std::string orig_name;
+  bool is_custom;
+};
+
+} // namespace
+
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    std::string message_name, const DescriptorProto& message, bool is_proto2,
```

Comment:
Can you thread in a FileDescriptor::Syntax instead of a bool for is_proto2?

Reply:
done